### PR TITLE
Support const-eval for `==` and `!=`

### PR DESCRIPTION
**Pull request by ike709**
I haven't tested every single comparison but basic testing shows that it works. This PR is required for one of the preproc tests to pass in https://github.com/wixoaGit/OpenDream/pull/827

### CODE REVIEW

File: DMCompiler/DM/Expressions/Constant.cs [hunk 1]
```diff
@@ -3,6 +3,7 @@
 using OpenDreamShared.Json;
 using System;
 using System.Collections.Generic;
+using Robust.Shared.Localization;
```
Comment (ike709) [inline DMCompiler/DM/Expressions/Constant.cs:6]:
```suggestion
```

Comment (Altoids1) [reply DMCompiler/DM/Expressions/Constant.cs]: what

File: DMCompiler/DM/Expressions/Constant.cs [hunk 2]
```diff
@@ -223,6 +248,22 @@ class Number : Constant {
...
+            return new Number(Location, 1); // true
+        }
+
+        public override Constant NotEquals(Constant rhs) {
+            if (rhs is not Number rhsNum || !MathHelper.CloseTo(rhsNum.Value, Value)) {
+                return new Number(Location, 1); // true
+            }
+
+            return new Number(Location, 0); // false
+        }
```
Comment (ike709) [inline DMCompiler/DM/Expressions/Constant.cs:266]:
```suggestion
        public override Constant Equals(Constant rhs) {
            if (rhs is not Number rhsNum || !MathHelper.CloseTo(rhsNum.Value, Value)) {
                return new Number(Location, 0); // false
            }

            return new Number(Location, 1); // true
        }

        public override Constant NotEquals(Constant rhs) {
            if (rhs is not Number rhsNum || MathHelper.CloseTo(rhsNum.Value, Value)) {
                return new Number(Location, 1); // true
            }

            return new Number(Location, 0); // false
        }
```

**Comment by Altoids1** [comment]
No longer a dependency of the preproc PR (but you should merge it anyways since Ike was very nice to do this for me)

File: DMCompiler/DM/Expressions/Binary.cs [hunk 3]
```diff
@@ -278,6 +288,16 @@ public NotEqual(Location location, DMExpression lhs, DMExpression rhs)
...
             proc.NotEqual();
         }
+
+        public override bool TryAsConstant(out Constant constant) {
+            if (!LHS.TryAsConstant(out var lhs) || !RHS.TryAsConstant(out var rhs)) {
+                constant = null;
+                return false;
+            }
+
+            constant = lhs.NotEquals(rhs);
```
Comment (wixoaGit) [inline DMCompiler/DM/Expressions/Binary.cs:298]: Would there ever a case where `NotEquals()` isn't just `!Equals()`?

**Review by wixoaGit — CHANGES_REQUESTED** [review]

File: DMCompiler/DM/Expressions/Binary.cs (hunk 3 above)
Comment (wixoaGit) [reply DMCompiler/DM/Expressions/Binary.cs:298]: Might just be able to use `IsTruthy()` for this too.

**Comment by ike709** [comment]
That's what I get for trying to webedit.

**Comment by Altoids1** [comment]
Eclipsed by #885.

**Pull request closed without merging**